[PM-23409] feat: Add client certificate authentication (mTLS) support for self-hosted environments

jalenfran opened the pull request:
## 🎟️ Tracking

- GitHub Discussion: [Add mTLS Support Discussion](https://github.com/orgs/bitwarden/discussions/15491)
- Feature Request: [Implementing mTLS in the Bitwarden apps](https://community.bitwarden.com/t/implementing-mtls-in-the-bitwarden-apps/63958)

## 📔 Objective

This PR implements client certificate authentication (mTLS) support for iOS app when connecting to self-hosted Bitwarden environments that require client certificates.

**Key Features:**
- PKCS#12 (.p12/.pfx) certificate import with password support
- Secure certificate storage independent of user login
- mTLS HTTP client integration for server authentication
- Certificate management UI integrated into self-hosted server configuration
- Comprehensive error handling and user feedback

**Technical Implementation:**
- `ClientCertificateConfiguration` model for certificate data and metadata
- `ClientCertificateService` for secure certificate management operations
- `CertificateHTTPClient` with URLSession delegate for mTLS authentication
- Global certificate storage using existing app settings infrastructure
- SwiftUI interface for certificate import, display, and removal

This enables users to authenticate with self-hosted Bitwarden servers that require client certificates for enhanced security.

## 📸 Screenshots

<img width="431" alt="Screenshot 2025-07-05 at 9 00 37 PM" src="https://github.com/user-attachments/assets/1df79456-4ca7-4a4e-8255-276c78e0af6a" />
<img width="429" alt="Screenshot 2025-07-05 at 9 13 28 PM" src="https://github.com/user-attachments/assets/24c0f3ff-f1c6-4c81-afb3-694d35991f43" />

## ⏰ Reminders before review

- [x] Contributor guidelines followed
- [x] All formatters and local linters executed and passed
- [x] Written new unit and / or integration tests where applicable
- [x] Protected functional changes with optionality (feature flags) - *N/A: Feature is opt-in via certificate import*
- [x] Used internationalization (i18n) for all UI strings
- [x] CI builds passed
- [x] Communicated to DevOps any deployment requirements - *N/A: No deployment changes needed*
- [x] Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

**Key Areas for Review:**
- 🔐 Security implementation of certificate storage and mTLS authentication
- 🎨 UI/UX integration with existing self-hosted configuration flow
- 📝 Error handling for various certificate import scenarios
- ⚡ Performance impact of certificate validation and HTTP client changes
- 🧪 Test coverage for certificate management workflows

**Files to Focus On:**
- `ClientCertificateService.swift` - Core certificate management logic
- `CertificateHTTPClient.swift` - mTLS HTTP client implementation
- `SelfHostedView.swift` - UI integration and user experience
- `StateService.swift` & `AppSettingsStore.swift` - Secure storage implementation

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

jalenfran committed:
mTLS support

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/bitwarden/ios?pullRequest=1720) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/bitwarden/ios?pullRequest=1720) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/bitwarden/ios?pullRequest=1720) it.</sub>

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

maxkpower commented:
Hey @jalenfran, thanks a lot for your PR! Please excuse the long silence, an automation issue unfortunately kept this ticket off our review board. We do want to support mTLS and will be reviewing the PR shortly.

edgenative commented:
Adding a bump here.

Would love to see this functionality landed

AvallaSD commented:
@maxkpower Is there any updates on this feature? Really looking forward to use this feature
@jalenfran May be they're waiting for you to resolve conflict?

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

jalenfran committed:
added back in clientCertificateService lost in merge

KuyomieKurama commented:
@maxkpower do you have any Updates on this.

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

edgenative commented:
@vvolkgang you were involved in https://github.com/bitwarden/android/pull/4486 over on the android side for this.  Any help you can offer here to move this one along...?  it's been stuck a while

gpalsson commented:
might someone with correct permission do the checks? It's been 3 months now and nothing is happening?

matt-livefront commented on BitwardenShared/Core/Platform/Services/CertificateHTTPClient.swift:
```diff
@@ -0,0 +1,165 @@
+            body: data,
+            requestID: request.requestID
+        )
+    }
```
🎨 Assuming these are the same, I think you can take advantage of the existing URLSession extension methods implemented in [URLSession+HTTPClient.swift](https://github.com/jalenfran/ios/blob/b459a2c90ee760a92753ae5d70fe68226b70c0b8/Networking/Sources/Networking/Extensions/URLSession%2BHTTPClient.swift)


```suggestion
    func download(from urlRequest: URLRequest) async throws -> URL {
        try await urlSession.download(from: urlRequest)
    }

    func send(_ request: HTTPRequest) async throws -> HTTPResponse {
        try await urlSession.send(request)
    }
```

matt-livefront commented on BitwardenShared/Core/Platform/Services/ClientCertificateService.swift:
```diff
@@ -0,0 +1,174 @@
+    private func getCertificateExpirationDate(_ certificate: SecCertificate) -> Date {
+        // For iOS, we'll use a default expiration date since extracting the actual date requires complex parsing
+        Date().addingTimeInterval(365 * 24 * 60 * 60) // 1 year from now
+    }
```
🎨 I don't think we'll end up showing these certificate values in the UI. And given that the issuer and expiration date are hardcoded here, we can remove this.

matt-livefront commented on BitwardenShared/Core/Platform/Models/Domain/ClientCertificateConfiguration.swift:
```diff
@@ -0,0 +1,67 @@
+    let certificateData: Data?
+
+    /// The certificate password.
+    let password: String?
```
⚠️ We should move the certificate and password to be stored in the Keychain, rather than UserDefaults. The Keychain is encrypted and has better access controls for this type of data.

matt-livefront commented on BitwardenShared/Core/Platform/Services/CertificateHTTPClient.swift:
```diff
@@ -0,0 +1,165 @@
+            var importResult: CFArray?
+            let importOptions: [String: Any] = [
+                kSecImportExportPassphrase as String: certificateInfo.password,
+            ]
```
🎨 I think it'd be great to move much of this logic into `ClientCertificateService` and expose a method which can return a `SecIdentity` for use in the `HTTPClient`. That should make this logic easier to test and keep all of the certificate handling/parsing together.

matt-livefront commented on BitwardenShared/Core/Platform/Services/ClientCertificateService.swift:
```diff
@@ -0,0 +1,174 @@
+        ]
+
+        var importResult: CFArray?
+        let status = SecPKCS12Import(data as CFData, importOptions as CFDictionary, &importResult)
```
🎨 One improvement I think you could make here. Once the certificate is imported with the password, the `SecIdentity` object (containing the private key and certificate) returned from `importResult` could be saved to the keychain. At that point, the password isn't needed in the future to decrypt the certificate, and it's one less thing to store and manage. And the certificate is still encrypted in the Keychain with the Keychain's encryption.

matt-livefront reviewed (changes requested):
Thanks for putting this together! This is off to a great start, but there are a few things that are still needed to get it merged in.

The big items would be storing the certificate in the Keychain, support for multiple users, and updating the UI to match the Android implementation.

We’d also want to make sure there’s full test coverage and all strings are set up for localizations.

Let us know if you’re willing to keep moving this effort forward!

jalenfran commented:
> @jalenfran What's the situation here? Are you still interested in implementing this? Or should someone else take over the correction?

Hi, I'm still a university student, and I have been a little busy on some course stuff and other projects recently. I will be working on a new PR soon.

jalenfran committed:
Merge branch 'bitwarden:main' into mTLS-support-iOS

jalenfran committed:
Refactor mTLS support for iOS: Implement app-level PKCS#12 import and per-user persistence

This commit addresses original PR feedback regarding iOS platform differences and security requirements for mutual TLS authentication. Unlike Android, third-party iOS apps are strictly sandboxed and cannot access client certificates installed via system MDM profiles.

Key architectural and security changes:
* Replaced System Picker with File Import: Removed obsolete system certificate picking logic and UI models (e.g. SystemCertificatePicker).
* App-Level Certificate Storage: Added native file selection for .p12 and .pfx files. We now use SecPKCS12Import to securely extract and store the SecIdentity directly in the Bitwarden app's local Keychain.
* Strengthened UI Security: Addressed PR feedback by ensuring we no longer store raw certificate data or the plaintext certificate password in application state.
* Certificate State Persistence: Added an `alias` attribute to `ClientCertificateConfiguration` to preserve the certificate reference in `StateService`. Added an `appeared` effect to the `SelfHostedProcessor` so the UI automatically reloads the saved configuration across app launches.
* Refined Error Handling: Explicitly mapped the `errSecAuthFailed` status code during PKCS#12 imports to throw `ClientCertificateError.invalidPassword`, providing users with accurate password failure dialogs rather than generic corruption errors.
* Localization & Cleanup: Added completely localized strings for removing certificates and incorrect passwords. Cleaned up obsolete Action, Effect, and State properties.

jalenfran committed:
refactor: Implement a circular dependency breaking mechanism for `ServerCommunicationConfigClient` and add HTTP redirect rejection to `CertificateHTTPClient`.

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

jalenfran committed:
fixed mistake in merging

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

jalenfran committed:
Remove uneeded code and fix formatting issues I introduced.

jalenfran committed:
refactor: remove `ClientCertificateConfiguration` and use string alias for client certificates.

jalenfran committed:
Update client certificate keychain attributes to include access group and protection levels.

jalenfran commented on BitwardenShared/Core/Auth/Services/KeychainRepository.swift:
```diff
@@ -75,6 +79,7 @@ enum KeychainItem: Equatable, KeychainStorageKeyPossessing {
              .deviceAuthKey,
              .deviceAuthKeyMetadata,
              .deviceKey,
+             .clientCertificateIdentity,
```
Good point I switched this to kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly to allow mTLS authentication during background sync (e.g., vault syncing) while the device is locked.

jalenfran committed:
Update Keychain accessibility for `clientCertificateIdentity` from `WhenUnlockedThisDeviceOnly` to `AfterFirstUnlockThisDeviceOnly`.

jalenfran committed:
Refactor: Simplify client certificate configuration and modernize mTLS networking

jalenfran commented on BitwardenShared/Core/Auth/Services/KeychainRepository.swift:
```diff
@@ -546,6 +577,76 @@ extension DefaultKeychainRepository {
+            throw KeychainServiceError.osStatusError(status)
+        }
+
+        return unsafeBitCast(identityRef, to: SecIdentity.self)
```
Done! I’ve refactored all SecIdentity operations in KeychainRepository to use the existing KeychainService abstractions (search, add, and delete).

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

jalenfran committed:
Fix: Fixed build errors caused from refactoring from new commits.  Secondly, cleaned up warnings.

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

matt-livefront committed:
Alphabetize properties and methods

matt-livefront commented on BitwardenShared/Core/Auth/Services/KeychainRepository+UserSessionTests.swift:
```diff
@@ -24,9 +23,6 @@ final class KeychainRepositoryUserSessionTests: BitwardenTestCase {
         subject = DefaultKeychainRepository(
-            appIDService: AppIDService(
-                appIDSettingsStore: appIDSettingsStore,
-            ),
```
⛏️ I think this need to be reverted for the tests to build.

jalenfran committed:
Update BitwardenShared/Core/Platform/Services/CertificateHTTPClient.swift

Co-authored-by: Matt Czech <matt@livefront.com>

jalenfran committed:
Fix for tests to build

jalenfran committed:
refactor: Switch mTLS certificate storage to fingerprint-based model for multi-user support

Replace per-user Keychain labels with SHA-256 fingerprint-based labels to allow multiple users
to securely share the same imported certificate. Certificates are now stored once in Keychain
and associated with users via per-user alias + fingerprint tracking in AppSettingsStore.
Implement ref-counted deletion logic to prevent breaking other users' access when one user logs out.
Add old certificate cleanup on re-import when users update their certificate. Includes comprehensive
test coverage for multi-user scenarios, orphaned state detection, and edge cases.

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

matt-livefront committed:
Update localizations

jalenfran committed:
refactor: store mTLS cert metadata in EnvironmentURLData via EnvironmentService

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

matt-livefront committed:
Clean up / revert unneeded changes

matt-livefront committed:
Update ClientCertificateService to AutoMockable

matt-livefront committed:
Add AuthRepository test

matt-livefront committed:
Remove SelfHostedState.keyHost

matt-livefront committed:
Update logic around loading and saving certificate

matt-livefront committed:
Remove unused ClientCertificateService methods

matt-livefront committed:
Remove NoRedirectSessionDelegate; add tests for CertificateHTTPClient

matt-livefront committed:
Remove EnvironmentService.clientCertificateAlias

matt-livefront commented on BitwardenShared/Core/Auth/Services/KeychainRepository.swift:
```diff
@@ -546,6 +579,68 @@ extension DefaultKeychainRepository {
+        try? keychainService.delete(query: deleteQuery as CFDictionary)
+
+        // 2. Add new
+        let addAttributes: [String: Any] = [
```
Thanks @jalenfran! I need to do some additional testing next week, but this is looking to be in a good spot.

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

matt-livefront committed:
Alphabetize ClientCertificateService

matt-livefront committed:
SelfHostedView alphabetizing

matt-livefront committed:
Add ErrorReporter to ClientCertificateService and SelfHostedProcessor

EuKeck commented:
What is the status of this feature? Is there still one "requested change" open or is the implementation finished? Thanks for the good work so far!

matt-livefront committed:
Add additional SelfHostedProcessor tests

matt-livefront committed:
Update KeychainRepository and add tests

matt-livefront committed:
Remove unused dialogShowPassword

matt-livefront committed:
Update EnvironmentService and add tests

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS

matt-livefront committed:
Update Authenticator

matt-livefront commented:
> > @jalenfran Just a heads up, I added a few commits to build on your changes here: mostly filling in some tests, aligning a few things with our conventions, and trimming a bit of extra functionality to keep things simpler.
> 
> Hi Matt, I was taking a look and I really appreciate the quickness and helping me get this through!

You bet! Thank you for all of your work on putting this PR together and working through the changes!

KatherineInCode commented on BitwardenShared/UI/Auth/Landing/SelfHosted/SelfHostedProcessor.swift:
```diff
@@ -14,9 +14,13 @@ protocol SelfHostedProcessorDelegate: AnyObject {
 final class SelfHostedProcessor: StateProcessor<SelfHostedState, SelfHostedAction, SelfHostedEffect> {
+    // MARK: Types
+
+    typealias Services = HasClientCertificateService & HasErrorReporter
```
🎨 I think this should be turned into two lines:

```suggestion
    typealias Services = HasClientCertificateService
        & HasErrorReporter
```

matt-livefront committed:
Update ClientCertificateServiceTests to use Swift Testing

matt-livefront committed:
Update SelfHostedProcessor services formatting

Co-authored-by: Katherine Bertelsen <katherine@katherineincode.com>

matt-livefront committed:
Update localizations

matt-livefront commented on BitwardenResources/Localizations/en.lproj/Localizable.strings:
```diff
@@ -1326,6 +1326,23 @@
+"ClientCertificateMtls" = "Client Certificate (mTLS)";
+"ACertificateWithTheAliasAlreadyExistsDescriptionLong" = "A certificate with the alias \"%1$@\" already exists. Do you want to replace it?";
+"EnterTheCertificatePasswordAndAlias" = "Enter the certificate password and the desired alias.";
+"ImportCertificate" = "Import Certificate";
```
Yes, good catch!

jalenfran committed:
Merge branch 'main' into mTLS-support-iOS
